### PR TITLE
Manage ORTModule configurations consistently

### PR DESCRIPTION
**Pull request by pengwa**
### Manage ORTModule options

Move all env vars that used for feature ON/OFF into runtime options for consistent managements. 

Be noted: the features' switch are assigned in 2 phases: default values, overwritten by env vars (if specified by users). So env vars take the highest priority when all 2 phases both given value explicitly for one feature.

### CODE REVIEW

File: docs/ORTModule_Training_Guidelines.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ More options for **developers**.
 +	from onnxruntime.training.ortmodule import ORTModule, DebugOptions, LogLevel
 +	model = ORTModule(model, DebugOptions(save_onnx=True, log_level=LogLevel.VERBOSE, onnx_prefix="model_name"))
 ```
-Check [DebugOptions implementation](../orttraining/orttraining/python/training/ortmodule/debug_options.py) for more details.
+Check [DebugOptions implementation](../orttraining/orttraining/python/training/ortmodule/_options.py) for more details.
```
Comment (baijumeswani) [inline docs/ORTModule_Training_Guidelines.md:50]: Shall we rename to `options.py` considering `DebugOptions` is public facing?

File: orttraining/orttraining/python/training/ortmodule/experimental/json_config/_load_config_from_json.py
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/experimental/json_config/_load_config_from_json.py:1]: The functionality to load config from a json file will no longer work after this change.

Shall we update this file to work with this centralized settings class?

File: orttraining/orttraining/python/training/ortmodule/_options.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+        return not _SkipCheck.is_disabled(self) and check in self
+
+    def is_disabled(self):
+        """Check whether `_SkipCheck.SKIP_CHECK_DISABLED is set on the `_SkipCheck instance"""
+
+        return _SkipCheck.SKIP_CHECK_DISABLED in self
+
+
+class _RuntimeOptions:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_options.py:158]: `RuntimeOptions` seems too broad a scope. It seems we have put all the configuration in this class? Do you think we could bucket similar configs into a group and then create a class for each of them?
They all can be a part of a larger `Options` class (including `DebugOptions`).

File: orttraining/orttraining/python/training/ortmodule/_options.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        #   to perform such operations in reduced (16-bit) precision.
+        # - If propagate_cast_ops_level is positive, 1 or 2, then in addition to opcode codes specified by
+        #   propagate_cast_ops_allow, use onnxruntime predetermined list of opcodes considered safe to move
+        #   before/after the cast operation.
+        # - Onnxruntime Level 1 predetermined "FP16 safe" opcodes include only opcodes that do not perform
+        #   any computation such as Transpose, Split, Reshape, etc., or the computation is actually in Float
+        #   such as GeLU, etc.
+        # - Whereas Level 2 predetermined "FP16 safe" opcodes include opcodes that perform computation using
+        #   contrib ops, Dropout, LayerNormalization, etc.
+        self._propagate_cast_ops_level = 1
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_options.py:190]: Considering the class is already internal (`_RuntimeOptions`), shall we name the members without the `_` prefixed?

File: orttraining/orttraining/python/training/ortmodule/_options.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+        # default execution order is priority-based for both dynamic/static shape input for now
+        # if we observe the benefit of static shape, we can expose this flag to the user
+        self._use_static_shape = False
+
+        # flag to enable symbolic shape inference for dynamic shape inputs to improve performance
+        self._run_symbolic_shape_infer = True
+
+        # PyTorch custom Autograd function support
+        from ._custom_autograd_function import custom_autograd_function_enabler
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_options.py:203]: This import will always be executed with ORTModule (since it is part of the constructor). Shall we move it to the top of this file?

File: orttraining/orttraining/python/training/ortmodule/experimental/json_config/_load_config_from_json.py
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/experimental/json_config/_load_config_from_json.py:1]: you are right, I forget about it.

File: orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
+# _hierarchical_ortmodule.py
 import tempfile
 import warnings
 
 import torch
 
 from .... import ortmodule
 from ... import ORTModule
-from ...debug_options import DebugOptions, LogLevel
+from ...options import DebugOptions, LogLevel
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py:11]: Possible to use absolute import?

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py [hunk 6]
```diff
@@ -87,6 +87,4 @@ def enable_custom_autograd_support(to_enable=True):
 from onnxruntime.training import ortmodule  # noqa: E402
 
 # Enable the custom autograd by default when PythonOp backend support is enabled during build.
-enable_custom_autograd_support(
-    not ortmodule._defined_from_envvar("ORTMODULE_DISABLE_CUSTOM_AUTOGRAD_SUPPORT", 0) and is_torch_interop_default_on()
-)
+enable_custom_autograd_support(is_torch_interop_default_on())
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py:90]: Maybe `is_torch_interop_default_on` should be imported at `options.py` as part of the options API instead of being a private property at `onnxruntime.capi._pybind_state` namespace?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 7]
```diff
@@ -89,6 +64,7 @@ def __init__(
 
         self._logger = logger
 
+        self._rt_options = RuntimeOptions(self._logger)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:67]:
```suggestion
        self._runtime_options = RuntimeOptions(self._logger)
```
Would prefer full words instead of abbreviations (rt -> runtime)

File: orttraining/orttraining/python/training/ortmodule/options.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+        self.onnx_opset_version = ortmodule.ONNX_OPSET_VERSION
+        self.conv_algo_search = "HEURISTIC"
+
+        # Configuration for cast optimization.
+        # Specify cast propagation strategy. Currently, three strategies are available:
+        #  NONE, INSERT-AND-REDUCE and FLOOD-FILL
+        # The default is FLOOD_FILL, expand FP16 computation regions in the graph using
+        # allowed opcodes for the given level.
+        self._propagate_cast_ops_strategy = C.PropagateCastOpsStrategy.FLOOD_FILL
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/options.py:178]:
```suggestion
        self.propagate_cast_ops_strategy = C.PropagateCastOpsStrategy.FLOOD_FILL
```
Several options in the class are starting with `_` which in python means "don't use outside this class".

File: orttraining/orttraining/python/training/ortmodule/options.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+                self.fallback_policy = _FallbackPolicy[self.fallback_policy]
+
+        # Configuration for skip check.
+        if "ORTMODULE_SKIPCHECK_POLICY" in os.environ:
+            self.skip_check = reduce(
+                lambda x, y: x | y,
+                [_SkipCheck[name] for name in parse_os_env_skip_check_flags("ORTMODULE_SKIPCHECK_POLICY")],
+            )
+
+    def get_feature_map(self) -> List[Tuple[str, bool, str]]:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/options.py:292]: Seems a separate feature from `Options`

**Review by thiagocrepaldi — COMMENTED** [review]
Overall, looks good. just a few minor comments

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py [hunk 10]
```diff
@@ -81,3 +81,12 @@ def enable_custom_autograd_support(to_enable=True):
             unregister_custom_op_symbolic("::prim_PythonOp", 1)
 
         custom_autograd_function_enabler.state = False
+
+
+from onnxruntime.capi._pybind_state import is_torch_interop_default_on  # noqa: E402
+from onnxruntime.training import ortmodule  # noqa: E402
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function.py]:
```suggestion
from onnxruntime.training import ortmodule 
```

move import to top if possible?

**Review by pengwa — COMMENTED** [review]
Thanks a lot for the view, update per your suggestions.

**Review by thiagocrepaldi — COMMENTED** [review]
LGTM

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 11]
```diff
@@ -102,44 +79,12 @@ def __init__(
...
-        #   as "FP16 safe", to insert/(re)move cast operations before/after to perform such operations in reduced (16-bit) precision.
-        # - If propagate_cast_ops_level is positive, 1 or 2, then in addition to opcode codes specified by propagate_cast_ops_allow, use onnxruntime
-        #   predetermined list of opcodes considered safe to move before/after the cast operation.
-        # - Onnxruntime Level 1 predetermined "FP16 safe" opcodes include only opcodes that do not perform any computation such as Transpose, Split, Reshape, etc.,
-        #   or the computation is actually in Float such as GeLU, etc.
-        #   whereas Level 2 predetermined "FP16 safe" opcodes include opcodes that perform computation using contrib ops, Dropout, LayerNormalization, etc.
-        self._propagate_cast_ops_level = 1
-        # List of opcodes to be considered safe to move before/after the cast operation if propagate_cast_ops_level is zero.
-        self._propagate_cast_ops_allow = []
+        self._rt_inspector = RuntimeInspector(self._logger)
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: Fixed. Thanks!

**Review by baijumeswani — APPROVED** [review]
Thanks for making this change.

**Review by mindest — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks @baijumeswani @thiagocrepaldi @justinchuby @mindest for the review . Merged with all green CIs. Feel free to let me know if there are more comments. I can cover later.

File: orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py (hunk 5 above)
Comment (justinchuby) [reply orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py]: The way the source code is laid out needs to match the import path. I think we can consider updating how the code is organized
